content/00.build/65.developer-reference/50.era-contracts/60.elliptic-curve-precompiles.md

```diff
@@ -0,0 +1,311 @@
+
+/// @notice Computes the Montgomery subtraction.
+/// @dev See <https://en.wikipedia.org/wiki/Montgomery_modular_multiplication#The_The_REDC_algorithm> for further details on the Montgomery multiplication.
+/// @param minuend The minuend in Montgomery form.
+/// @param subtrahend The subtrahend in Montgomery form.
+/// @return ret The result of the Montgomery subtraction.
+function montgomerySub(minuend, subtrahend) -> ret {
+    ret := montgomeryAdd(minuend, sub(P(), subtrahend))
+}
+
+```
+
+We do not use `addmod`. That's because in most cases the sum does not exceed the modulus.
+
+### Montgomery Multiplication
+
+The product of $aR \mod N$ and $bR \mod N$ is $REDC((aR \mod N)(bR \mod N))$.
+
+```solidity
+/// @notice Computes the Montgomery multiplication using the Montgomery reduction algorithm (REDC).
+/// @dev See <https://en.wikipedia.org/wiki/Montgomery_modular_multiplication#The_The_REDC_algorithm> for further details on the Montgomery multiplication.
+/// @param multiplicand The multiplicand in Montgomery form.
+/// @param multiplier The multiplier in Montgomery form.
+/// @return ret The result of the Montgomery multiplication.
+function montgomeryMul(multiplicand, multiplier) -> ret {
+    let hi := getHighestHalfOfMultiplication(multiplicand, multiplier)
+    let lo := mul(multiplicand, multiplier)
+    ret := REDC(lo, hi)
+}
+
+```
+
+### Montgomery Inversion
+
+```solidity
+/// @notice Computes the Montgomery modular inverse skipping the Montgomery reduction step.
+/// @dev The Montgomery reduction step is kept because a modification in the binary extended Euclidean algorithm is used to compute the modular inverse.
+/// @dev See the function `binaryExtendedEuclideanAlgorithm` for further details.
+/// @param a The field element in Montgomery form to compute the modular inverse of.
+/// @return invmod The result of the Montgomery modular inverse (in Montgomery form).
+function montgomeryModularInverse(a) -> invmod {
+    invmod := binaryExtendedEuclideanAlgorithm(a)
+}
+```
+
+As said before, we use a modified version of the bEE algorithm that lets us “skip” the Montgomery reduction step.
+
+The regular algorithm would be $REDC((aR \mod N)^{−1}(R^3 \mod N))$ which involves a regular inversion plus a multiplication by a value that can be precomputed.
+
+## ECADD
+
+Precompile for computing elliptic curve point addition. The points are represented in affine form, given by a pair of coordinates $(x,y)$.
+
+Affine coordinates are the conventional way of expressing elliptic curve points, which use 2 coordinates. The math is concise and easy to follow.
+
+For a pair of constants $a$ and $b$, an elliptic curve is defined by the set of all points $(x,y)$
+that satisfy the equation $y^2=x^3+ax+b$, plus a special “point at infinity” named $O$.
+
+### Point Doubling
+
+To compute $2P$ (or $P+P$), there are three cases:
+
+- If $P = O$, then $2P = O$.
+- Else $P = (x, y)$
+  - If $y = 0$, then $2P = O$.
+  - Else $y≠0$, then
+
+        $$
```

Comment:
I don't think having tabs before LaTeX (`$$`) is supported:
![image](https://github.com/matter-labs/zksync-docs/assets/1890113/0f891675-5ab7-4167-b510-7e5e0af94ccd)

Reply:
Yeah, we should remove it then